pkg/coreapi/graph/loaders/trace.go

```diff
@@ -0,0 +1,726 @@
+	mod.StepOp = &stepOp
+	mod.Duration = &dur
+	mod.Status = status
+	mod.StepInfo = models.WaitForEventStepInfo{
+		EventName:    evtName,
+		Expression:   expr,
+		Timeout:      timeout,
+		FoundEventID: foundEvtID,
+		TimedOut:     expired,
+	}
+
+	// TODO: output
+
+	tb.processed[span.SpanID] = true
+	tb.processed[wait.SpanID] = true
+
+	return nil
+}
+
+func (tb *TraceTreeBuilder) processInvoke(ctx context.Context, span *cqrs.Span, mod *models.RunTraceSpan) error {
+	// invoke span always have a nested span that stores the details of the invoke
+	if len(span.Children) != 1 {
+		return fmt.Errorf("missing invoke details")
+	}
+	invoke := span.Children[0]
+
+	stepOp := models.StepOpInvoke
+	var (
+		runID         *ulid.ULID
+		returnEventID *ulid.ULID
+		timedOut      *bool
+	)
+
+	// timeout
+	expstr, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeExpires]
+	if !ok {
+		return fmt.Errorf("missing invoke expiration time")
+	}
+	exp, err := strconv.ParseInt(expstr, 10, 64)
+	if err != nil {
+		return fmt.Errorf("error parsing expiration timestamp")
+	}
+	timeout := time.UnixMilli(exp)
+
+	// triggering event ID
+	evtIDstr, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeTriggeringEventID]
+	if !ok {
+		return fmt.Errorf("missing invoke triggering event ID")
+	}
+	triggeringEventID, err := ulid.Parse(evtIDstr)
+	if err != nil {
+		return fmt.Errorf("error parsing invoke triggering event ID: %w", err)
+	}
+
+	// target function ID
+	fnID, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeTargetFnID]
+	if !ok {
+		return fmt.Errorf("missing invoke target function ID for invoke")
+	}
+
+	// run ID
+	if str, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeRunID]; ok {
+		id, err := ulid.Parse(str)
+		if err != nil {
+			return fmt.Errorf("error parsing invoke run ID: %w", err)
+		}
+		runID = &id
+	}
+
+	// return event ID
+	if str, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeReturnedEventID]; ok {
+		evtID, err := ulid.Parse(str)
+		if err != nil {
+			return fmt.Errorf("error parsing invoke return event ID: %w", err)
+		}
+		returnEventID = &evtID
+
+		exp := false
+		timedOut = &exp
+	}
+
+	// final timed out check
+	if timeout.Before(time.Now()) {
+		var exp bool
+		if str, ok := invoke.SpanAttributes[consts.OtelSysStepInvokeExpired]; ok {
+			exp, _ = strconv.ParseBool(str)
+		}
+		timedOut = &exp
+	}
+
+	// set invoke details
+	mod.StepOp = &stepOp
+	mod.StepInfo = models.InvokeStepInfo{
+		TriggeringEventID: triggeringEventID,
+		FunctionID:        fnID,
+		RunID:             runID,
+		ReturnEventID:     returnEventID,
+		Timeout:           timeout,
+		TimedOut:          timedOut,
+	}
+
+	if returnEventID != nil {
+		status := models.RunTraceSpanStatusCompleted
+		if invoke.StatusCode == "STATUS_CODE_ERROR" {
+			status = models.RunTraceSpanStatusFailed
+		}
+		mod.Status = status
+	} else {
+		if timedOut != nil && *timedOut {
+			mod.Status = models.RunTraceSpanStatusFailed
+		}
+	}
+
+	// TODO: output
```

Comment:
Later PR?

Reply:
Yup, output and trigger APIs are separate PRs. this PR itself is getting big.